lib/src/model.dart

```diff
@@ -4596,6 +4888,32 @@ class PackageGraph {
     return hrefMap;
   }
 
+  void _collectLibraryNames(library) {
+    void collectClassNames(Class c) {
+      nameRegistry.add(c.package);
+      if (!c._mixins.isEmpty) {
+        c._mixins.forEach((t) {
+          c.nameRegistry.add(t.element);
+        });
+      }
+      if (c._supertype != null) {
+        c.nameRegistry.add(c._supertype.element);
+      }
+      if (!c.interfaces.isEmpty) {
+        c.interfaces.forEach((t) {
+          c.nameRegistry.add(t.element);
+        });
+      }
+    }
+
+    nameRegistry.add(library);
+    library._allModelElements
```

Comment:
I think only canonical ModelElements are likely to be relevant here, as those are the only ones we will ever generate links to.  However, doing all ModelElements the same might simplify matters so OK to leave this if it does.

Reply:
Also, this is literally _all_ modelElements for the library, including nested members of classes, parameters, and so on that will never be written into the same directory.  That might cause "collisions" that aren't real.